runtime/src/append_vec.rs

```diff
@@ -702,11 +702,13 @@ pub mod tests {
     }
 
     impl AppendVecStoredAccountMeta<'_> {
-        #[allow(clippy::cast_ref_to_mut)]
         fn set_data_len_unsafe(&self, new_data_len: u64) {
             // UNSAFE: cast away & (= const ref) to &mut to force to mutate append-only (=read-only) AppendVec
             unsafe {
-                *(&self.meta.data_len as *const u64 as *mut u64) = new_data_len;
+                std::ptr::write(
+                    std::mem::transmute::<*const u64, *mut u64>(&self.meta.data_len),
+                    new_data_len,
+                );
```

Comment:
This is still UB. Casting/transmuting a immutable reference to a mutable reference is always UB. The only sound way of mutating an immutable reference to a mutable reference is to use a [`Cell`](https://doc.rust-lang.org/std/cell/struct.Cell.html) or [`UnsafeCell`](https://doc.rust-lang.org/std/cell/struct.UnsafeCell.html).

See this [playground example](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=c1f740e1b970d2444f153c9c008d259e) for proof with Miri.
```
error: Undefined Behavior: attempting a write access using <1679> at alloc852[0x0], but that tag only grants SharedReadOnly permission for this location
  --> src/main.rs:7:9
   |
7  | /         std::ptr::write(
8  | |             std::mem::transmute::<*const u64, *mut u64>(&data_len),
9  | |             new_data_len,
10 | |         );
   | |         ^
   | |         |
   | |_________attempting a write access using <1679> at alloc852[0x0], but that tag only grants SharedReadOnly permission for this location
   |           this error occurs as part of an access at alloc852[0x0..0x8]
   |
   = help: this indicates a potential bug in the program: it performed an invalid operation, but the Stacked Borrows rules it violated are still experimental
   = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
help: <1679> was created by a SharedReadOnly retag at offsets [0x0..0x8]
  --> src/main.rs:8:57
   |
8  |             std::mem::transmute::<*const u64, *mut u64>(&data_len),
   |                                                         ^^^^^^^^^
   = note: BACKTRACE (of the first span):
   = note: inside `main` at src/main.rs:7:9: 10:10
```

NOTE: The current `cast_ref_to_mut` or the now uplifted `invalid_reference_casting` lint unfortunately do not (yet) lint against, but I'm planning on fixing it.